101.md

```diff
@@ -0,0 +1,111 @@
+# NIP-101: Alias Exchange
+
+## Abstract
+
+This NIP proposes the 'Alias Exchange' protocol, designed to facilitate the secure exchange of alias keys in private communication, thus protecting users' real npub keys. The protocol includes four kinds of events, each serving a specific function in the exchange process.
+
+## Motivation
+
+The Alias Exchange protocol enhances the security of NIP04 (private chat) by providing a means to protect real npub keys. By exchanging alias keys, users can communicate without revealing their real keys, reducing the risk of unauthorized access and potential data breaches. This protocol can be particularly useful in scenarios such as friend requests, acceptance of friend requests, rejection of friend requests, and removal of friends.
+
+
+## Specification
+
+### Kinds
+
+- **10100 - Request Exchange**: This event is initiated by the user who wants to start the alias key exchange. The initiator generates an alias public key derived from their own private key and the recipient's public key. The event sent is signed with the alias private key, and 'p' is the recipient's real public key.
+
+- **10101 - Accept Exchange**: This event represents the acceptance of the request for the alias key exchange. The recipient generates their own alias key using the same algorithm. The 'p' in this event is the initiator's alias public key.
+
+- **10102 - Reject Exchange**: This event signifies the rejection of the request for the alias key exchange. The rejection event is signed with the recipient's alias and sent to the initiator.
+
+- **10103 - Remove Alias**: This event is used to remove the alias. It is signed with the initiator's alias and sent to the recipient's alias public key.
+
+### Implementation
+
+The protocol can be implemented following the steps below. Note that the content of the content field in this protocol is encrypted following the NIP04 protocol:
+
+1. User A (the initiator) generates an alias key pair derived from their own private key and User B's (the recipient) public key.
+
+2. User A sends a 10100 kind event to User B, signed with their alias private key. The 'p' parameter is User B's real pubkey. The `encrypted_text` can be in json format: `{"p":"A's real pubkey", "content":"your content"}`.
```

Comment:
Shouldn't the encrypted_text contain some form of verification that A is actually sending this message? Any attacker could just put A's pubkey into the event and B would have no way to tell if it was actually sent by A.

Reply:
Thank you, you're absolutely right. We can add a "sig" field in the text:

`{
    "p":"A's real pubkey",
    "content":"your content", 
    "sig":"<64-bytes hex of the signature of the event id, signed by A's real private key>"
}`

Reply:
What is the "event id" in this scenario? We only have p and content fields, is it just some hash of that?

Reply:
Apologies for any confusion earlier. Let me clarify: it's the request event id send to B, we will have two signatures. 

The first signature of the event id is generated with A's alias private key and is included in the `sig` field of the event(NIP01)

The second signature of the event id is generated with A's actual private key and is included within the encrypted `content`. 

Here's how the event would look:

```json
{
 "id": "event id",
 "pubkey": "A's alias public key for B",
 "kind": 10100,
 "tags": [
    ["p", "B's real public key"],
  ],
 "content": "<encrypted_text>?iv=<initialization_vector>",
 "sig": "signature generated with A's alias private key",
 ...other fields
}
```
In the content field, <encrypted_text> would be a JSON string:
```json
{ 
  "p":"A's real pubkey", 
  "content":"your content",  
  "sig":"<The second signature , signed by A's real private key>" 
}
```

Reply:
Apologies for any confusion earlier. Let me clarify: 

data: `[eventCreateTime, p, content]` , where eventCreateTime  corresponds to the time the event was created
sig: 64-bytes hex of the signature of the sha256 hash of the `data` mentioned above, generated with A's real private key.